File: Python/Arrays/L_findKthPairDiff.py

def smallestDistancePair(nums, k: int) -> int:
    def countPairs(distance: int) -> int:
        count = left = 0
        for right in range(1, len(nums)):
            while nums[right] - nums[left] > distance:
                left += 1
            count += right - left
        return count

    nums.sort()
    low, high = 0, nums[-1] - nums[0]
    
    while low < high:
        mid = (low + high) // 2
        if countPairs(mid) < k:
            low = mid + 1
        else:
            high = mid
    return low


print(smallestDistancePair([1,6,1], 3))

File: JavaScript/Array/L_findKthPairDiff.js
var smallestDistancePair = function (nums, k) {
    nums.sort((a, b) => a - b);
    let low = 0;
    let high = nums[nums.length - 1] - nums[0];

    while (low < high) {
        let mid = Math.floor((low + high) / 2);
        if (countPairs(nums, mid) < k) {
            low = mid + 1;
        } else {
            high = mid;
        }
    }

    return low;
};

function countPairs(nums, distance) {
    let count = 0;
    let left = 0;
    for (let right = 1; right < nums.length; right++) {
        while (nums[right] - nums[left] > distance) {
            left++;
        }
        count += right - left;
    }
    return count;
}
